src/main/java/io/reactivex/internal/operators/single/SingleCache.java

```diff
@@ -13,97 +13,166 @@
-    protected void subscribeActual(SingleObserver<? super T> s) {
-
-        Object o = notification.get();
-        if (o != null) {
-            s.onSubscribe(EmptyDisposable.INSTANCE);
-            if (NotificationLite.isError(o)) {
-                s.onError(NotificationLite.getError(o));
+    protected void subscribeActual(final SingleObserver<? super T> s) {
+        CacheDisposable<T> d = new CacheDisposable<T>(s, this);
+        s.onSubscribe(d);
+        
+        if (add(d)) {
+            if (d.isDisposed()) {
+                remove(d);
+            }
+        } else {
+            Throwable ex = error;
+            if (ex != null) {
+                s.onError(ex);
             } else {
-                s.onSuccess(NotificationLite.<T>getValue(o));
+                s.onSuccess(value);
             }
             return;
         }
         
-        synchronized (subscribers) {
-            o = notification.get();
-            if (o == null) {
-                subscribers.add(s);
+        if (wip.getAndIncrement() == 0) {
+            source.subscribe(this);
+        }
+    }
+
+    boolean add(CacheDisposable<T> observer) {
+        for (;;) {
+            CacheDisposable<T>[] a = observers.get();
+            if (a == TERMINATED) {
+                return false;
+            }
+            int n = a.length;
+            @SuppressWarnings("unchecked")
+            CacheDisposable<T>[] b = new CacheDisposable[n + 1];
+            System.arraycopy(a, 0, b, 0, n);
+            b[n] = observer;
+            if (observers.compareAndSet(a, b)) {
+                return true;
             }
         }
-        if (o != null) {
-            s.onSubscribe(EmptyDisposable.INSTANCE);
-            if (NotificationLite.isError(o)) {
-                s.onError(NotificationLite.getError(o));
+    }
+    
+    @SuppressWarnings("unchecked")
+    void remove(CacheDisposable<T> observer) {
+        for (;;) {
+            CacheDisposable<T>[] a = observers.get();
+            int n = a.length;
+            if (n == 0) {
+                return;
+            }
+            
+            int j = -1;
+            for (int i = 0; i < n; i++) {
+                if (a[i] == observer) {
+                    j = i;
+                    break;
+                }
+            }
+            
+            if (j < 0) {
+                return;
+            }
+            
+            CacheDisposable<T>[] b;
+            
+            if (n == 1) {
+                b = EMPTY;
             } else {
-                s.onSuccess(NotificationLite.<T>getValue(o));
+                b = new CacheDisposable[n - 1];
+                System.arraycopy(a, 0, b, 0, j);
+                System.arraycopy(a, j + 1, b, j, n - j - 1);
+            }
+            if (observers.compareAndSet(a, b)) {
+                return;
             }
-            return;
         }
+    }
+
+    @Override
+    public void onSubscribe(Disposable d) {
+        // not supported by this operator
```

Comment:
Ah. I don't think that's worth it. I just misunderstood the comment and thought "not allowed" instead of "not supported".